superduperdb/components/model.py

```diff
@@ -90,361 +85,364 @@ def get(self, k, default=None):
+            select = Serializable.from_dict(select)
 
-    def _forward(
-        self, X: t.Sequence[int], num_workers: int = 0, **kwargs
-    ) -> t.Sequence[int]:
-        if self.batch_predict:
-            return self.to_call(X, **kwargs)
+        if validation_sets:
+            from superduperdb.components.dataset import Dataset
 
-        outputs = []
-        if num_workers:
-            to_call = _to_call(self.to_call, **kwargs)
-            pool = multiprocessing.Pool(processes=num_workers)
-            for r in pool.map(to_call, X):
-                outputs.append(r)
-            pool.close()
-            pool.join()
+            validation_sets = list(validation_sets)
+            for i, vs in enumerate(validation_sets):
+                if isinstance(vs, Dataset):
+                    assert db is not None
+                    db.add(vs)
+                    validation_sets[i] = vs
+
+        self.training_configuration = configuration or self.training_configuration
+
+        if db is not None:
+            db.add(self)
+
+        if db is not None and db.compute.type == 'distributed':
+            return self.create_fit_job(
+                X,
+                select=select,
+                y=y,
+                **kwargs,
+            )(db=db, dependencies=dependencies)
         else:
-            for r in X:
-                outputs.append(self.to_call(r, **kwargs))
-        return outputs
+            return self._fit(
+                X,
+                y=y,
+                configuration=configuration,
+                data_prefetch=data_prefetch,
+                db=db,
+                metrics=metrics,
+                select=select,
+                validation_sets=validation_sets,
+                **kwargs,
+            )
 
-    def _predict(self, X: t.Any, one: bool = False, **predict_kwargs):
-        if one:
-            return self._predict_one(X)
+    def append_metrics(self, d: t.Dict[str, float]) -> None:
+        if self.metric_values is not None:
+            for k, v in d.items():
+                self.metric_values.setdefault(k, []).append(v)
 
-        if self.preprocess:
-            preprocessed_X = []
-            for r in X:
-                r = self.setup_required_inputs(r)
-                preprocessed_X.append(self.preprocess(**r))
-            X = preprocessed_X
 
-        elif self.preprocess is not None:
-            raise ValueError('Bad preprocess')
-        if self.collate_fn:
-            X = self.collate_fn(X)
-        outputs = self._forward(X, **predict_kwargs)
+@wraps(_TrainingConfiguration)
+def TrainingConfiguration(identifier: str, **kwargs):
+    return _TrainingConfiguration(identifier=identifier, kwargs=kwargs)
+
 
-        if self.postprocess:
-            outputs = [self.postprocess(o) for o in outputs]
-        elif self.postprocess is not None:
-            raise ValueError('Bad postprocess')
+@dc.dataclass
+class Signature:
+    singleton: t.ClassVar[str] = 'singleton'
```

Comment:
Allows developers to have different types of input to their `model.predict(...)`: